entity-service-impl/src/main/java/org/hypertrace/entity/data/service/EntityDataServiceImpl.java

```diff
@@ -148,19 +164,26 @@ public void getAndUpsertEntities(Entities request, StreamObserver<Entity> respon
     try {
 
       Map<Key, Document> documentMap = new HashMap<>();
+      List<Entity> updatedEntities = new ArrayList<>();
       for (Entity entity : request.getEntityList()) {
         Entity normalizedEntity = this.entityNormalizer.normalize(tenantId, entity);
+        updatedEntities.add(normalizedEntity);
         Document doc = convertEntityToDocument(normalizedEntity);
         Key key = this.entityNormalizer.getEntityDocKey(tenantId, normalizedEntity);
         documentMap.put(key, doc);
       }
 
-      Streams.stream(entitiesCollection.bulkUpsertAndReturnOlderDocuments(documentMap))
-          .flatMap(document -> PARSER.<Entity>parseOrLog(document, Entity.newBuilder()).stream())
-          .map(Entity::toBuilder)
-          .map(builder -> builder.setTenantId(tenantId))
-          .map(Entity.Builder::build)
-          .forEach(responseObserver::onNext);
+      List<Entity> existingEntities =
+          Streams.stream(entitiesCollection.bulkUpsertAndReturnOlderDocuments(documentMap))
+              .flatMap(
+                  document -> PARSER.<Entity>parseOrLog(document, Entity.newBuilder()).stream())
+              .map(Entity::toBuilder)
+              .map(builder -> builder.setTenantId(tenantId))
+              .map(Builder::build)
+              .collect(Collectors.toList());
+
+      sendChangeNotification(tenantId, existingEntities, updatedEntities);
```

Comment:
most of our notifications happen after we notify the response observer, this one happens before. We should be consistent (after, IMO).

Reply:
Done